SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc

```diff
@@ -0,0 +1,967 @@
+  subdetEfficiencies_(conf_specific),
+  flatDistribution_((addNoise || AddPixelInefficiency || fluctuateCharge || addThresholdSmearing) ? new CLHEP::RandFlat(eng, 0., 1.) : 0),
+  gaussDistribution_((addNoise || AddPixelInefficiency || fluctuateCharge || addThresholdSmearing) ? new CLHEP::RandGaussQ(eng, 0., theReadoutNoise) : 0),
+  // Threshold smearing with gaussian distribution:
+  smearedThreshold_Endcap_(addThresholdSmearing ? new CLHEP::RandGaussQ(eng, theThresholdInE_Endcap , theThresholdSmearing_Endcap) : 0),
+  smearedThreshold_Barrel_(addThresholdSmearing ? new CLHEP::RandGaussQ(eng, theThresholdInE_Barrel , theThresholdSmearing_Barrel) : 0),
+  rengine_(&eng)
+{
+  LogInfo("Phase2TrackerDigitizerAlgorithm") << "Phase2TrackerDigitizerAlgorithm constructed\n"
+			    << "Configuration parameters:\n"
+			    << "Threshold/Gain = "
+			    << "threshold in electron Endcap = "
+			    << theThresholdInE_Endcap
+			    << "\nthreshold in electron Barrel = "
+			    << theThresholdInE_Barrel
+			    << " " << theElectronPerADC << " " << theAdcFullScale
+			    << " The delta cut-off is set to " << tMax
+			    << " pix-inefficiency " << AddPixelInefficiency;
+}
+
+Phase2TrackerDigitizerAlgorithm::~Phase2TrackerDigitizerAlgorithm() {
+  LogDebug("Phase2TrackerDigitizerAlgorithm") << "Phase2TrackerDigitizerAlgorithm deleted";
+}
+
+Phase2TrackerDigitizerAlgorithm::SubdetEfficiencies::SubdetEfficiencies(const edm::ParameterSet& conf) {
+  barrel_efficiencies = conf.getParameter< std::vector<double> >("EfficiencyFactors_Barrel");
+  endcap_efficiencies = conf.getParameter< std::vector<double> >("EfficiencyFactors_Endcap");
+}
+// =================================================================
+//
+// Generate primary ionization along the track segment.
+// Divide the track into small sub-segments
+//
+// =================================================================
+void Phase2TrackerDigitizerAlgorithm::primary_ionization(const PSimHit& hit, 
+							 std::vector<DigitizerUtility::EnergyDepositUnit>& ionization_points) const {
+  // Straight line approximation for trajectory inside active media
+  const float SegmentLength = 0.0010; // in cm (10 microns)
+  float energy;
+
+  // Get the 3D segment direction vector
+  LocalVector direction = hit.exitPoint() - hit.entryPoint();
+
+  float eLoss = hit.energyLoss();  // Eloss in GeV
+  float length = direction.mag();  // Track length in Silicon
+
+  int NumberOfSegments = int (length / SegmentLength); // Number of segments
+  if (NumberOfSegments < 1) NumberOfSegments = 1;
+  LogDebug("Phase2TrackerDigitizerAlgorithm")
+    << "enter primary_ionzation " << NumberOfSegments
+    << " shift = "
+    << hit.exitPoint().x() - hit.entryPoint().x() << " "
+    << hit.exitPoint().y() - hit.entryPoint().y() << " "
+    << hit.exitPoint().z() - hit.entryPoint().z() << " "
+    << hit.particleType() 
+    << " " << hit.pabs();
+
+  float* elossVector = new float[NumberOfSegments];  // Eloss vector
```

Comment:
even better an dyn-array on the stack
float elossVector[NumberOfSegments];  // Eloss vector